systemtest/loadgen/eventhandler/handler.go

```diff
@@ -20,120 +20,106 @@ package eventhandler
+	Limiter   *rate.Limiter
+}
+
 // New creates a new tracehandler.Handler from a glob expression, a filesystem,
 // Transport and an optional rate limitter. If the rate limitter is empty, an
 // infinite rate limitter will be used. and The glob expression must match one
 // file containing one or more batches.
 // The file contents should be in plain text, but the in-memory representation
 // will use zlib compression (BestCompression) to avoid compressing the batches
 // while benchmarking and optimizing memory usage.
-func New(p string, t *Transport, storage fs.FS, l *rate.Limiter) (*Handler, error) {
-	if t == nil {
+func New(config Config) (*Handler, error) {
+	if config.Transport == nil {
 		return nil, errors.New("empty transport received")
 	}
-
-	var buf bytes.Buffer
-	zw, err := zlib.NewWriterLevel(&buf, zlib.BestCompression)
-	if err != nil {
-		return nil, err
-	}
-	writer := compressedWriter{
-		buf:     &buf,
-		zwriter: zw,
-	}
-	if l == nil {
-		l = rate.NewLimiter(rate.Inf, 0)
+	if config.Limiter == nil {
+		config.Limiter = rate.NewLimiter(rate.Inf, 0)
 	}
+
 	h := Handler{
-		transport:  t,
-		limiter:    l,
-		readerPool: newByteReaderPool(),
+		config: config,
+		writerPool: sync.Pool{
+			New: func() any {
+				pw := &pooledWriter{}
+				zw, err := zlib.NewWriterLevel(&pw.buf, zlib.BestSpeed)
+				if err != nil {
+					panic(err)
+				}
+				pw.Writer = zw
+				return pw
+			},
+		},
 	}
-	matches, err := fs.Glob(storage, p)
+	matches, err := fs.Glob(config.Storage, config.Path)
 	if err != nil {
 		return nil, err
 	}
 	for _, path := range matches {
-		f, err := storage.Open(path)
+		f, err := config.Storage.Open(path)
 		if err != nil {
 			return nil, err
 		}
+
 		s := bufio.NewScanner(f)
-		var scanned uint
+		var current *batch
 		for s.Scan() {
 			line := s.Bytes()
 			if len(line) == 0 {
 				continue
 			}
+
 			// TODO(marclop): Suppport RUM headers and handle them differently.
 			if bytes.HasPrefix(line, rumMetaHeader) {
 				return nil, errors.New("rum data support not implemented")
 			}
 
-			if isMeta := bytes.HasPrefix(line, metaHeader); !isMeta {
-				writer.WriteLine(line)
-				scanned++
-				continue
-			}
-
-			// Since the current token is a metadata line, it means that we've
-			// read the start of the batch, because of that, we'll only flush,
-			// close and copy the compressed contents when the scanned events
-			// are greater than 0. The first iteration will only write the meta
-			// and continue decoding events.
-			// After writing the whole batch to the buffer, we reset it and the
-			// zlib.Writer so it can be used for the next iteration.
-			if scanned > 0 {
-				if err := writer.Close(); err != nil {
-					return nil, err
-				}
-				h.batches = append(h.batches, batch{
-					bytes: writer.Bytes(),
-					items: scanned,
-				})
+			// Copy the line, as it will be overwritten by the next scan.
+			linecopy := make([]byte, len(line))
+			copy(linecopy, line)
+			line = linecopy
```

Comment:
Why is this assignment necessary? It seems that we're copying `line` to `linecopy`, but `copy()` doesn't zero the `line` variable, right?

Reply:
It's not, I had intended to use `line` everywhere below, but I didn't :smile:
I'll remove this.